src/CharRNN/index.js

```diff
@@ -0,0 +1,245 @@
+          this.cellsAmount += 1;
+        } else {
+          this.model[`Bias_${key.match(/[0-9]/)[0]}`] = vars[key];
+        }
+      } else if (key.match(regexFullyConnected)) {
+        if (key.match(regexWeights)) {
+          this.model.fullyConnectedWeights = vars[key];
+        } else {
+          this.model.fullyConnectedBiases = vars[key];
+        }
+      } else {
+        this.model[key] = vars[key];
+      }
+    });
+    await this.loadVocab(path);
+    await this.initCells();
+    return this;
+  }
+
+  async loadVocab(path) {
+    const json = await fetch(`${path}/vocab.json`)
+      .then(response => response.json())
+      .catch(err => console.error(err));
+    this.vocab = json;
+    this.vocabSize = Object.keys(json).length;
+  }
+
+  async initCells() {
+    this.cells = [];
+    this.zeroState = { c: [], h: [] };
+    const forgetBias = tf.tensor(1.0);
+
+    const lstm = (i) => {
+      const cell = (DATA, C, H) =>
+        tf.basicLSTMCell(forgetBias, this.model[`Kernel_${i}`], this.model[`Bias_${i}`], DATA, C, H);
+      return cell;
+    };
+
+    for (let i = 0; i < this.cellsAmount; i += 1) {
+      this.zeroState.c.push(tf.zeros([1, this.model[`Bias_${i}`].shape[0] / 4]));
+      this.zeroState.h.push(tf.zeros([1, this.model[`Bias_${i}`].shape[0] / 4]));
+      this.cells.push(lstm(i));
+    }
+
+    this.state = this.zeroState;
+  }
+
+
+  async generateInternal(options) {
+    await this.ready;
+    const seed = options.seed || this.defaults.seed;
+    const length = +options.length || this.defaults.length;
+    const temperature = +options.temperature || this.defaults.temperature;
+    const stateful = options.stateful || this.defaults.stateful;
+    if (!stateful) {
+      this.state = this.zeroState;
+    }
+
+    const results = [];
+    const userInput = Array.from(seed);
+    const encodedInput = [];
+
+    userInput.forEach((char) => {
+      encodedInput.push(this.vocab[char]);
+    });
+
+    let input = encodedInput[0];
+    let probabilitiesNormalized = []; // will contain final probabilities (normalized)
+
+    for (let i = 0; i < userInput.length + length + -1; i += 1) {
+      const onehotBuffer = tf.buffer([1, this.vocabSize]);
+      onehotBuffer.set(1.0, 0, input);
+      const onehot = onehotBuffer.toTensor();
+      let output;
+      if (this.model.embedding) {
+        const embedded = tf.matMul(onehot, this.model.embedding);
+        output = tf.multiRNNCell(this.cells, embedded, this.state.c, this.state.h);
+      } else {
+        output = tf.multiRNNCell(this.cells, onehot, this.state.c, this.state.h);
+      }
+
+      this.state.c = output[0];
+      this.state.h = output[1];
+
+      const outputH = this.state.h[1];
+      const weightedResult = tf.matMul(outputH, this.model.fullyConnectedWeights);
+      const logits = tf.add(weightedResult, this.model.fullyConnectedBiases);
+      const divided = tf.div(logits, tf.tensor(temperature));
```

Comment:
A temperature of 0 is invalid, should we protect against that here or leave it up to the end user?

Reply:
This should fix it `const temperature = temp > 0 ? temp : 0.1;`